content/blog/2024/oci-manifest-capability.md

```diff
@@ -0,0 +1,227 @@
+more detailed inspections and validations, aligning with security standards and
+organizational protocols.
+
+Some possible uses cases to this new capability is allow policies to inspects the
+platform supported by the given container images. As well as inspecting annotations,
+user, environment variable and many other information available inside the image
+manifests.
+
+### Getting Started
+
+To showcase this capability, let's create a sample policy that fetches and logs
+the OCI image manifest. Consider this Rust validation code:
+
+```rust 
+use lazy_static::lazy_static;
+
+use guest::prelude::*;
+use kubewarden_policy_sdk::wapc_guest as guest;
+
+extern crate kubewarden_policy_sdk as kubewarden;
+use kubewarden::{logging, protocol_version_guest, request::ValidationRequest, validate_settings};
+use kubewarden::host_capabilities::oci::manifest;
+
+mod settings;
+use settings::Settings;
+
+use slog::{info, o, warn, Logger};
+
+lazy_static! {
+    static ref LOG_DRAIN: Logger = Logger::root(
+        logging::KubewardenDrain::new(),
+        o!("policy" => "sample-policy")
+    );
+}
+
+#[no_mangle]
+pub extern "C" fn wapc_init() {
+    register_function("validate", validate);
+    register_function("validate_settings", validate_settings::<Settings>);
+    register_function("protocol_version", protocol_version_guest);
+}
+
+fn validate(payload: &[u8]) -> CallResult {
+    let validation_request: ValidationRequest<Settings> = ValidationRequest::new(payload)?;
+
+    match validation_request.extract_pod_spec_from_object() {
+        Ok(pod_spec) => {
+            pod_spec.unwrap_or_default().containers.iter().for_each(|container| {
+                match  manifest(&container.image.clone().unwrap_or_default()) {
+                    Ok(manifest) => {
+                        let manifest_json = serde_json::to_string(&manifest).unwrap_or_default();
+                        info!(LOG_DRAIN, "{}", manifest_json);
+                    }
+                    Err(error) => {
+                        warn!(LOG_DRAIN, "cannot fetch container image manifest: {:?}", error);
+
+                    }
+                }
+
+            })
+        },
+        Err(_) => {
+            warn!(LOG_DRAIN, "cannot unmarshal resource: this policy does not know how to evaluate this resource; accept it");
+        }
+    }
+
+    kubewarden::accept_request()
+}
+```
+
+This straightforward code retrieves the pod spec, iterates over the containers,
+and logs the OCI image manifest for each container.
+
+Now, observe the output using test data from the Rust policy template:
+
+```console 
+$ jq ".object.spec.containers" test_data/pod_creation.json
+[
+  {
+    "image": "nginx",
+    "name": "nginx"
+  },
+  {
+    "image": "ghcr.io/kubewarden/policies/psp-user-group:v0.1.2",
+    "name": "kubewarden policy"
+  }
+]
+
+$ kwctl run  --request-path test_data/pod_creation.json  annotated-policy.wasm 
```

Comment:
This output is too long, we're losing interest from the reader. Why don't you change the policy to just print out the architecture of the images being referenced?

Reply:
Done